go-client/ring.go

```diff
@@ -0,0 +1,218 @@
+			MaxConnAge:         time.Minute * 10, // all connections cannot live longer than this
+			IdleCheckFrequency: time.Minute * 5,  // reap staled connections after 5 minutes
+		},
+	}
+	defaultOpts.Addresses = addr
+	return NewRing(defaultOpts)
+}
+
+// NewRing return a Ring behaves like a single client, with consistent hashing
+func NewRing(opts RingOptions) (*Ring, error) {
+	r := &Ring{
+		opts: opts,
+		mu:   sync.RWMutex{},
+	}
+	switch opts.HashAlgorithm {
+	case Rendezvous:
+		r.resetConsistentHash = func(addrs []string) ConsistentHash {
+			return newRendezvous(addrs)
+		}
+	default:
+		return nil, fmt.Errorf("unsupported hash algorithm: %s", opts.HashAlgorithm)
+	}
+	allShards := make(map[string]*shardClient)
+	clientOpt := opts.ClientOptions
+	for _, addr := range opts.Addresses {
+		clientOpt := ClientOptions{
+			Pool: pool.Options{
+				Address:            addr,
+				PoolTimeout:        clientOpt.PoolTimeout,
+				PoolSize:           clientOpt.PoolSize,
+				MinIdleConn:        clientOpt.MinIdleConn,
+				Dialer:             clientOpt.Dialer,
+				IdleTimeout:        clientOpt.IdleTimeout,
+				MaxConnAge:         clientOpt.MaxConnAge,
+				IdleCheckFrequency: clientOpt.IdleCheckFrequency,
+			},
+		}
+		client, err := NewClient(clientOpt)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create client for address %s: %s", addr, err)
+		}
+		allShards[addr] = &shardClient{
+			downCount: 0,
+			client:    client,
+		}
+	}
+	r.allShards = allShards
+	r.updateAliveShards()
+	go r.healthCheckByInterval(opts.HealthCheckFrequency)
+	r.commandAdaptor = r.cmdWithConsistentHash
+	return r, nil
+}
+
+func (r *Ring) Close() error {
+	return nil
+	//TODO
+}
+
+func (r *Ring) cmdWithConsistentHash(cmd Cmd) error {
+	r.mu.RLock()
+	args := cmd.Args()
+	if len(args) == 0 {
+		return fmt.Errorf("command with no argument is not supported by ring just yet")
+	}
+	addr := r.consistentHash.Lookup(args[0])
+	fmt.Printf("arg: %s hashed into: %s\n", args[0], addr)
+	correctShard, ok := r.allShards[addr]
+	r.mu.RUnlock()
+	if !ok {
+		return fmt.Errorf("address after consistent hash %s does not exist", addr)
+	}
+	return correctShard.client.commandAdaptor(cmd)
+}
+
+func (r *Ring) getShards() []*shardClient {
+	var ret []*shardClient
+	r.mu.RLock()
+	for _, s := range r.allShards {
+		ret = append(ret, s)
+	}
+	r.mu.RUnlock()
+	return ret
+}
+
+func (r *Ring) healthCheckByInterval(frequency time.Duration) {
+	ticker := time.NewTicker(frequency)
+	defer ticker.Stop()
+	for range ticker.C {
+		var shardsChanged bool
+
+		for _, shard := range r.getShards() {
+			err := shard.client.Ping()
+			isUp := err == nil || err == pool.ErrPoolTimeout
+			statusChanged := shard.updateStatus(isUp)
+			if statusChanged {
+				shardsChanged = true
+			}
+		}
+		if shardsChanged {
+			r.updateAliveShards()
+		}
+	}
+}
+
+func (r *Ring) updateAliveShards() {
+	r.mu.RLock()
```

Comment:
Should move `r.mu.RUnlock()` below `RLock()`, and should you `defer r.mu.RUnlock()`